keps/sig-apps/20190412-pdb-for-scale-subresource.md

```diff
@@ -0,0 +1,84 @@
+
+
+## Summary
+
+[Pod Disruption Budget (PDB)](https://kubernetes.io/docs/tasks/run-application/configure-pdb/)
+is a Kubernetes API that limits the number of pods of a collection that are down simultaneously from voluntary disruptions. PDBs allows a user to specify the allowed disruption through either min available or max unavailable number of pods. In order to support PDBs where max number of unavailable pods is set, the PDB controller needs to be able to look up the desired number of replicas. It does this by looking at the controller(s) (as specified by the owner ref) for the pods covered by the PDB. Currently only four of the basic workload controllers are supported for PDBs, namely Deployment, StatefulSet, ReplicaSet and ReplicaController. 
+
+The scale subresource allow any resource to specify its desired number of replicas and a generic way to look up this information. This document lays out a plan to use the scale subresource to allow setting PDBs for any resource implementing the scale subresource.
+
+## Motivation
+
+PDBs are an important tool to control the number of voluntary disruptions for workloads on Kubernetes. As more users start deploying custom controllers/operators based on CRDs, it is inconvenient that PDBs are not available. The scale subresource is already in use by the autoscaler and we should extend its use to PDBs.
+
+### Goals
+
+- Implement support for scale subresources in PDBs
+- Avoid major performance impact from the change
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+In the current implementation, the PDB controller identifies the workload controller by going through all the pods covered by the PDB, and for each pod, check in sequence for each of the four workload controllers. The PDB controller then looks up the desired number of replicas from the identified controller. If the set of pods covered by the PDB identfies more than one controller, that is considered an error. 
+
+The plan is to keep this approach, but check for the scale subresource if neither of the kubernetes workload controllers match. Each of the kubernetes workload controllers actually implement the scale subresource, so it is possible to only use that one. But since shared informers can not be used with the scale subresource, we want to rely on the informers for checking the kubernetes workload controllers, and only try the scale subresource if none of the them match.
+
+As mentioned above, the PDB controller will look up the controller for every pod covered by the PDB. In almost all scenarios the controller will be the same for all pods and therefore the desired number of replicas will be the same. This is not a concern for the kubernetes workload controllers since the controller uses shared informers. When using the scale subresource, this would lead to unnecessary calls to the apiserver. We will implement a per reconcile loop cache, so the PDB controller only needs to look up each controller once per reconcile loop.
```

Comment:
I'm very confused about why it would do this.

If it literally just needs to map pod -> desired scale, it should load the controller ref, map that object reference to a rest path (client-go should be able to do this for you), read the /scale subresource.

You can avoid overly many queries by caching the responses for a limited amount of time. (URL request path should be a sufficient cache key.)

I don't think watching really gains very much in this case, and it's complicated to set up because in the general case it needs to start and stop watcher dynamically.

Reply:
This is in line with what I am suggesting. Only difference is that instead of using a cache based on time, it will cache the desired scale for each controller per reconcile loop. So for every loop, each controller will only be looked up once, even if there is a large number of pods with the same controller.
I have updated the KEP to make this clearer.

Reply:
OK, this is fine. Time based is probably even better (since there's no guarantee on how long a reconcile loop will take), but we can always change it later.